tools/ttysnoop.py

```diff
@@ -62,33 +69,75 @@ def usage():
 bpf_text = """
 #include <uapi/linux/ptrace.h>
 #include <linux/fs.h>
+#include <linux/uio.h>
 
-#define BUFSIZE 256
+#define BUFSIZE USER_DATASIZE
 struct data_t {
     int count;
     char buf[BUFSIZE];
 };
 
+BPF_ARRAY(data_map, struct data_t, 1);
 BPF_PERF_OUTPUT(events);
 
+static int do_tty_write(void *ctx, const char __user *buf, size_t count)
+{
+    int zero = 0, i;
+    struct data_t *data;
+
+    data = data_map.lookup(&zero);
+    if (!data)
+        return 0;
+
+    for (i = 0; i < USER_DATACOUNT; i++) {
+        bpf_probe_read_user(&data->buf, BUFSIZE, (void *)buf);
+        if (count > BUFSIZE)
+            data->count = BUFSIZE;
+        else
+            data->count = count;
+        events.perf_submit(ctx, data, sizeof(struct data_t));
+        if (count < BUFSIZE)
+            return 0;
+        count -= BUFSIZE;
+    }
+
+    return 0;
+};
+
+/**
+ * commit 9bb48c82aced (v5.11-rc4) tty: implement write_iter
+ * changed arguments of tty_write function
+ */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 11, 0)
 int kprobe__tty_write(struct pt_regs *ctx, struct file *file,
     const char __user *buf, size_t count)
 {
     if (file->f_inode->i_ino != PTS)
         return 0;
 
-    // bpf_probe_read_user() can only use a fixed size, so truncate to count
-    // in user space:
-    struct data_t data = {};
-    bpf_probe_read_user(&data.buf, BUFSIZE, (void *)buf);
-    if (count > BUFSIZE)
-        data.count = BUFSIZE;
-    else
-        data.count = count;
-    events.perf_submit(ctx, &data, sizeof(data));
+    return do_tty_write(ctx, buf, count);
+}
+#else
+int kprobe__tty_write(struct pt_regs *ctx, struct kiocb *iocb,
+                      struct iov_iter *from)
+{
+    const char __user *buf;
+    struct kvec *kvec;
+    size_t count;
 
-    return 0;
-};
+    if (iocb->ki_filp->f_inode->i_ino != PTS)
+        return 0;
+
+    if (from->type != (ITER_IOVEC + WRITE))
+        return 0;
+
+    kvec  = from->kvec;
+    buf   = kvec->iov_base;
+    count = kvec->iov_len;
```

Comment:
buf and count are not used.

Reply:
leftover from trampoline removal, fixed, thanks